Fixes #34504 - Add support to template_kind_id when importing templates

adiabramovitch opened the pull request:
DRAFT 
Fixed a bug I found when adding the support, added a few tests to ensure valid params.

adiabramovitch committed:
Fixes #34504 - Add support to template_kind_id when importing provisioning template

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -84,13 +84,16 @@ def ignore_default
   # Set attributes based on template +text and metadata found in this +text
   # the metadata parsing can be adjusted using +options
-  def import_without_save(text, options = {})
+  def import_without_save(name, text, options = {})
```
You are changing a method signature ( adding parameter), what has big consequences on ALL the callers, not just one and I believe this method is called from mutiple places.
Given it is just for validation of the name, that is not great.

adiabramovitch replied:
You are right, that is why it is a draft. 
I'll find a different way :)

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -99,14 +102,23 @@ def import_without_save(text, options = {})
 
+  def valid_params?(name, metadata)
+    if !metadata["kind"].nil? && (metadata["snippet"] == true)
+      raise Foreman::Exception.new(N_('snippet and template kind don\'t match'))
```
That is not very informative message, it doesn't tell me much about what exactly I should fix, does it?

adiabramovitch replied:
sure, do you have a better suggestion? :)

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -84,13 +84,16 @@ def ignore_default
+      self.snippet = !!@importing_metadata["snippet"]
+      self.default = options[:default] unless options[:default].nil?
+      self.description = @importing_metadata["description"]
+      self.template_kind_id = @importing_metadata["kind"]
```
This is not correct per my assumption. Shouldn't metadata be a human readable name of the template kind and template_kind_id be an inner representation? This will not get translated automatically, we will need to look for the kind by its name.

adiabramovitch replied:
> Hi, the RM ticket nor the PR description explains what you're trying to fix here so it is hard to review. I'll assume that you want to import the kind from template file as `template_kind_id`.

Thanks for your comment. 
I updated the issue so it will be clearer.

ezr-ondrej reviewed (commented):
Hi, the RM ticket nor the PR description explains what you're trying to fix here so it is hard to review.
I'll assume that you want to import the kind from template file as `template_kind_id`.

adiabramovitch commented:
@ezr-ondrej I think I fixed the issue, can you take a look? :)
1. Is that ok to change the tests for the ptable and report? there is no metadata in those tests. 
2. Do you think I should add a rabl as well? 
I think that the failing test is not related to what I have changed

ezr-ondrej commented on test/controllers/api/v2/provisioning_templates_controller_test.rb:
```diff
@@ -136,9 +136,10 @@ class Api::V2::ProvisioningTemplatesControllerTest < ActionController::TestCase
+    show_response = ActiveSupport::JSON.decode(@response.body)
     assert_response :success
-    assert_match 'bbbb', ProvisioningTemplate.unscoped.find_by_name(snippet.name).template
+    assert_match show_response["template"], ProvisioningTemplate.unscoped.find_by_name(snippet.name).template
```
it is not a good assert, you are asserting whether it renders properly the result, but you do not assert what is the actual template being imported.
Could we keep these untouched and have another test that would test only the metadata?
That adds to my point - the name validation is making the life harder for users of the import.

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -99,6 +101,16 @@ def import_without_save(text, options = {})
+    if (!metadata["kind"].nil? && metadata["kind"] != "snippet") && (metadata["snippet"] == true)
+      raise Foreman::Exception.new(N_('When snippet is set as true, template kind shouldn\'t be present'))
+    end
+    unless name.eql?(metadata["name"])
```
I'm not quite sure what is this validation good for.
The template is looked for by a name, when can it fail so this would be actually useful?
This is hard to check and I do not see a value of it actually :thinking:
Also it forces the metadata in the template, which is not great, makes it harder for custom templates to be imported.
What is the usecase for this?

adiabramovitch replied:
Yes, I agree with you. It is a bit problematic when it forces the metadata. 
When I talked with @lzap about it, he suggested the name validation. 
It is supposed to ensure that the name corresponds to the metadata if exists. maybe that is the change I should do. 
@lzap Do you still think it is a good idea? :)

ezr-ondrej replied:
I believe we do have that validation sort of, the template is looked for by name. Could we implement the `template_kind` thing in this PR and possibly do the name validation separetly if we still feel it is important? It would make us move forward and simplify this PR

lzap replied:
This is absolutely not important, I just was just throwing few ideas what we could possibly do without thinking it through.

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -88,9 +88,11 @@ def import_without_save(text, options = {})
     self.default = options[:default] unless options[:default].nil?
-    self.description = @importing_metadata[:description]
+    self.description = @importing_metadata["description"]
+    template_kind = TemplateKind.find_by_name(@importing_metadata["kind"])
```
Find by name should not be used if not necessary :) Also if the kind is not set or is a snippet, I'd prefer if we do not query the database - it is bit more efficient.
```suggestion
    template_kind = TemplateKind.find_by(name: @importing_metadata["kind"]) if @importing_metadata["kind"] && @importing_metadata["kind"] != 'snippet'
```
might look better not to inline the `if`, but up to you :)

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -126,8 +137,11 @@ def self.parse_metadata(text)
   def self.import!(name, text, options = {})
-    template = import_without_save(name, text, options)
-    return template unless template.valid?
+    template = import_without_save(name, text, options) if valid_params?(parse_metadata(text))
```
Now I believe the check can be moved into `import_without_save` as we do not need to validate name and we will be able to parse metadata just once (remember that parsing text might be computationaly expensive operation :) )

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -126,8 +137,11 @@ def self.parse_metadata(text)
-    template = import_without_save(name, text, options)
-    return template unless template.valid?
+    template = import_without_save(name, text, options) if valid_params?(parse_metadata(text))
+    if template.valid?
+      template.save!
+      return template
+    end
```
Why we are addin these? It would seem that the lines below got it covered already.

adiabramovitch replied:
the validation was wrong from what I saw . 
It was: 
` return template unless template.valid?`
and the template wasn't saved.

ezr-ondrej replied:
isn't it correct? In case the template is invalid, we should not try to save it because it would fail anyway, or am I missing something?

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -99,6 +103,13 @@ def import_without_save(text, options = {})
     self
   end
 
+  def self.valid_params?(metadata)
```
would it be better to call this method `valid_metadata?`?
Also methods having `?` suffix should return `true/false`, your method does not do that, so in that light you might want to call the method `validate_metadata` (omiting the `?` suffix)

ezr-ondrej commented on app/models/template.rb:
```diff
@@ -88,9 +88,13 @@ def import_without_save(text, options = {})
     @importing_metadata = self.class.parse_metadata(text)
     Foreman::Logging.logger('app').debug "setting attributes for #{name} with id: #{id || 'N/A'}"
-    self.snippet = !!@importing_metadata[:snippet]
+    self.snippet = !!@importing_metadata["snippet"]
```
nitpick: given you decided to use string key, I'd prefer single quote strings as those are bit more efficient to evaluate:
```suggestion
    self.snippet = !!@importing_metadata['snippet']
```

adiabramovitch replied:
so I am not sure If I have decided that.
when I check the `@importing_metadata` output the keys appear as a string and not as a symbol. Is there a reason why?

ezr-ondrej replied:
Because it is parsed as [`ActiveSupport::HashWithIndifferentAccess`](https://api.rubyonrails.org/classes/ActiveSupport/HashWithIndifferentAccess.html)

ezr-ondrej reviewed (commented):
Few comments

ofedoren reviewed (commented):
Just a quick question, why do you want to set the template kind here if it's already being done here: https://github.com/theforeman/foreman/blob/cc089d3d8843a7e96137e296bf2d91a760c13d9e/app/models/provisioning_template.rb#L246.

If it's more about that the `template_kind_id` is missing (which is weird if so) then the same `import_custom_data` should be adjusted, shouldn't it?

UPD: Also, please, be aware that changing `template` model/contoller will affect not only `provisioning_templates`, but others as well.
